man/man5/rlm_realm.5
.\"     # DS - begin display
.de DS
.RS
.nf
.sp
..
.\"     # DE - end display
.de DE
.fi
.RE
.sp
..
.TH rlm_realm 5 "14 March 2004" "" "FreeRADIUS Module"
.SH NAME
rlm_realm \- FreeRADIUS Module
.SH DESCRIPTION
The \fIrlm_realm\fP module parses the User-Name attribute into a
User section and a Realm section.  This is used primarily in a
proxy situation, however, Realms can also be used locally to provide
different service profiles based on the Realm being used.
.PP
The main configuration items to be aware of are:
.IP format
This can be either 'prefix' or 'suffix'.  It specifies whether the
Realm is before or after the User portion in the User-Name string.
.IP delimiter
A single character in quotes, which is used as the delimiting
character that separates the Realm and User sections of the string.
.IP ignore_default
This is set to either 'yes' or 'no'.  If set to 'yes', this will 
prevent the module instance from matching a realm against the DEFAULT
entry.  This may be useful if you have multiple realm module instances.
The default is 'no'.
.IP ignore_null
This is set to either 'yes' or 'no'.  If set to 'yes', this will 
prevent the module instance from matching a realm against the NULL
entry.  This may be useful if you have multiple realm module instances.
The default is 'no'.
.PP
This module parses the realm from the User-Name attribute according
to the instance configuration, and then performs a lookup to find a
matching realm in the '/etc/raddb/proxy.conf' file.  Depending on the
configuration of the Realm as matched in the file, the username may
be rewritten in a 'stripped' format, or with the Realm portion
removed.  In either case, a Realm attribute is created and added to
the packet on a match, which can be used by other modules.
.SH CONFIGURATION
.PP
.DS
modules {
  ... stuff here ...
.br
.br
  # useranme@realm syntax
.br
  realm suffix {
.br
    format = suffix
.br
    delimiter = "@"
.br
  }
.br
.br
   # realm/username syntax
.br
   realm prefix {
.br
    format = prefix
.br
    delimiter = "/"
.br
  }
.br
.br
  ... stuff here ...
.br
}
.DE
.PP
.SH SECTIONS
.BR authorization,
.BR pre-accounting
.PP
.SH FILES
.I /etc/raddb/radiusd.conf,
.I /etc/raddb/proxy.conf
.PP
.SH "SEE ALSO"
.BR radiusd (8),
.BR radiusd.conf (5),
.BR proxy.conf (5)
.SH AUTHORS
Chris Parker, cparker@segv.org


doc/ChangeLog
FreeRADIUS 3.0.22 Tue 24 Mar 2020 12:00:00 EDT urgency=low
	Feature improvements
	* Update dictionary.bskyb.  Fixes #3339.
	* Update alcatel.sr. #3495
	* Abfab-tls updates from Alejandro Perez.
	* Allow CoA and Disconnect messages over TLS sockets.
	  Fixes #3356.
	* New xlat for setting status of rlm_always instances and new
	  resource-check example virtual server for manipulating control flow
	  in unlang policies based on status of some external resource.
	  Patches from Terry Burton.
	* Allow delta CRLs.  Fixes #3393.
	* Limited support for dynamic home servers.  See proxy.conf
	  and doc/configuration/dynamic_home_servers.md
	* Update radmin to show more information about the home servers
	  using "show home_server list all".
	* Support TLS 1.3 in TLS-based EAP methods.  This has been tested
	  to work with hostap and Windows.  Thanks to Jorge Vergara
	  for assistance.
	* The default configuration now replies with EAP-Key-Name, if it
	  is available, and was requested.
	* Include extensions in generated certificates.
	* Many improvements for DHCP IP pools for all SQL dialects.
	  Patches from Nick Porter.
	* Ignore user-provided dhparams in FIPS mode.
	  Patch from Alexander Scheel.
	* Add support for DHCP-Decline.
	* Remove native support for Cisco LEAP.  It is insecure, and
	  should not be used.  Proxying LEAP is still supported.
	* Automatically set fragment size / MTU, so that PEAP/EAP-TLS
	  works, and no longer requires manual changes to the configuration.

	Bug fixes
	* Respect the "log_reject" configuration item in more places.
	  Fixes #3352.  This lowers the number of "Login incorrect"
	  messages when "log_reject = no".
	* Fix rpmbuild for Centos > 6.  Patch from Matthew Newton.
	* Run Post-Proxy-Type Fail... when all home servers are down.
	* Note that rlm_replicate can only use UDP, and not TCP or TLS.
	* DHCP pool lookup is now keyed by Client Identifier (Option 61) when
	  supplied by client, otherwise the hardware address is used. Compliant
	  with RFC 2132. This change will not affect existing systems on upgrade,
	  but new installations will use the new behavior.
	  Patch from Terry Burton.
	* Fix minor spelling mistakes in man pages.  Patch from
	  Alexander Scheel.
	* Don't print invalid tags in rlm_cache, among other places.
	  Fixes #3491
	* Do home_server failover immediately when an initial TCP / Radsec
	  connection fails.
	* Port EAP-PWD constant time fixes from "master" branch.  The issue
	  was verified by Mohamed Sabt, and a patch supplied by
	  Daniel De Almeida Braga.
	* Clear error on SQLITE_BUSY to prevent memory leak in corner cases.
	  Patch from Nick Porter.
	* Properly add SQL clients to virtual servers.
	* Update documentation for cert generation.  Patch from Alexander Scheel.
	* Use better API when decoding DHCP packets, to avoid unnecessary work.
	  This improves performance noticably.

FreeRADIUS 3.0.21 Tue 24 Mar 2020 12:00:00 EDT urgency=low
	Feature improvements
	* New stored procedure for allocating IPs with PostgreSQL.
	  Rates of 1500 IPs per second are now possible.
	  See raddb/mods-config/sql/ippool/postgresql/procedure.sql
	  Patch from Terry Burton.
	* Add SQL IP pool support for Microsoft SQL Server
	  See raddb/mods-config/sql/ippool/mssql/
	  Patch from Terry Burton.
	* Added RCNTEC dictionary. Closes #3168.
	* Added Pica8 dictionary. Closes #3179.
	* Add TLS-Client-Cert-Valid-Since attribute holding notBefore date.
	  Patch from Boris Lytochkin.  Fixes #3157.
	* Generate attributes containing unknown OIDs.
	  See raddb/sites-available/tls.  Patch from Boris Lytochkin.
	* Update the WiMAX dictionary.
	* Added ability to rlm_python(Python2) show a stacktrace from errors. #2979
	* Add WiFi Alliance Policy OIDs.  See raddb/certs/xpextensions
	  Patch from Stefan Winter.
	* radmin now shows coa stats, too.
	* Sample schema extensions for summarizing data in SQL.
	  See mods-config/sql/main/*/process-radacct.sql
	  Many patches from Terry Burton.
	* Update dictionary.aerohive, dictionary.fortinet,
	  dictionary.arista and dictionary.erx
	* Added VAS Experts dictionary.
	* Many updates to RPM and jenkins builds from Matthew Newton
	* Added %C (time now in seconds) and %c (microsecond component of now)
	  back-ported from the "master" branch.
	* Add reload capability to systemd unit file in Debian and RedHat.
	* Increase timestamp precision in postauth to maximum supported by each
	  database and simplify (and make more consistent between drivers) the
	  timestamps in SQL queries by using expansions.
	  Patches from Terry Burton.
	* Option to set dictionary path in raduat script. Patch from Terry Burton.

	Bug fixes
	* Various fixes found by PVS-Studio.
	* Set permissions of certificates in bootstrap shell script.
	  Fixes #3132.
	* Increase the 'nasportid' SQL field for 'varchar(32)'. #3141
	* Skip processing proxy reply if there are no home servers
	  available.
	* Update SQLite IPPool queries.  Fixes #3177
	  Patch from Terry Burton.
	* rlm_sql_unixodbc fixes. Patches from Terry Burton. Fixes #2822
	* Fixes when building with LibreSSL. Patch from Nathan Owens.
	* Fix the rlm_python3 build. Note that this module is experimental. #3183
	* The rlm_python should append the 'python_path' paths in 'sys.path',
	  It fixes the expected behavior to use the existing Python modules.
	  Fixes #3180
	* Fix rlm_python to print the script errors properly.
	* Bound total query time for PostgreSQL.  Fixes #3253
	* Many fixes to Oracle sqlippool.  It now does 500 IPs per second
	  without any tuning.  Fixes #3270.
	* Reference sqlippool by it's correct name.  Fixes #3272
	* Revert 3.0.20 patch which caused crashes on duplicate clients.
	* Update WiMAX-MSK attribute.  Fixes #3280.
	* Fix crash when trying to access non-existant regex capture group.
	* Use timestamps (request or server) rather than SQL NOW() in
	  accounting queries so that these are stable when replayed from a
	  file buffer.
	  Patches from Terry Burton.


src/modules/rlm_realm/rlm_realm.c
/*
 *   This program is is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or (at
 *   your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
 */

/**
 * $Id$
 * @file rlm_realm.c
 * @brief Parses NAIs and assigns requests to realms.
 *
 * @copyright 2000-2013  The FreeRADIUS server project
 */
RCSID("$Id$")

#include <freeradius-devel/radiusd.h>
#include <freeradius-devel/modules.h>

#include "trustrouter.h"

#define  REALM_FORMAT_PREFIX   0
#define  REALM_FORMAT_SUFFIX   1

typedef struct rlm_realm_t {
	int		format;
	char const	*format_string;
	char const	*delim;
	bool		ignore_default;
	bool		ignore_null;

#ifdef HAVE_TRUST_ROUTER_TR_DH_H
	char const	*default_community;
	char const	*rp_realm;
	char const	*trust_router;
	uint32_t	tr_port;
	bool		rekey_enabled;
	uint32_t	realm_lifetime;
#endif
} rlm_realm_t;

static CONF_PARSER module_config[] = {
	{ "format", FR_CONF_OFFSET(PW_TYPE_STRING, rlm_realm_t, format_string), "suffix" },
	{ "delimiter", FR_CONF_OFFSET(PW_TYPE_STRING, rlm_realm_t, delim), "@" },
	{ "ignore_default", FR_CONF_OFFSET(PW_TYPE_BOOLEAN, rlm_realm_t, ignore_default), "no" },
	{ "ignore_null", FR_CONF_OFFSET(PW_TYPE_BOOLEAN, rlm_realm_t, ignore_null), "no" },
#ifdef HAVE_TRUST_ROUTER_TR_DH_H
	{ "default_community", FR_CONF_OFFSET(PW_TYPE_STRING, rlm_realm_t, default_community),  "none" },
	{ "rp_realm", FR_CONF_OFFSET(PW_TYPE_STRING, rlm_realm_t, rp_realm),  "none" },
	{ "trust_router", FR_CONF_OFFSET(PW_TYPE_STRING, rlm_realm_t, trust_router),  "none" },
	{ "tr_port", FR_CONF_OFFSET(PW_TYPE_INTEGER, rlm_realm_t, tr_port),  "0" },
	{ "rekey_enabled", FR_CONF_OFFSET(PW_TYPE_BOOLEAN, rlm_realm_t, rekey_enabled),  "no" },
	{ "realm_lifetime", FR_CONF_OFFSET(PW_TYPE_INTEGER, rlm_realm_t, realm_lifetime),  "0" },
#endif
	CONF_PARSER_TERMINATOR
};

/*
 *	Internal function to cut down on duplicated code.
 *
 *	Returns -1 on failure, 0 on no failure.  returnrealm
 *	is NULL on don't proxy, realm otherwise.
 */
static int check_for_realm(void *instance, REQUEST *request, REALM **returnrealm)
{
	char *namebuf;
	char *username;
	char const *realmname = NULL;
	char *ptr;
	VALUE_PAIR *vp;
	REALM *realm;

	struct rlm_realm_t *inst = instance;

	/* initiate returnrealm */
	*returnrealm = NULL;

	/*
	 *	If the request has a proxy entry, then it's a proxy
	 *	reply, and we're walking through the module list again.
	 *
	 *	In that case, don't bother trying to proxy the request
	 *	again.
	 *
	 *	Also, if there's no User-Name attribute, we can't
	 *	proxy it, either.
	 */
	if ((!request->username)
#ifdef WITH_PROXY
	    || (request->proxy != NULL)
#endif
	    ) {

		RDEBUG2("Proxy reply, or no User-Name.  Ignoring");
		return RLM_MODULE_NOOP;
	}

	/*
	 *      Check for 'Realm' attribute.  If it exists, then we've proxied
	 *      it already ( via another rlm_realm instance ) and should return.
	 */

	if (fr_pair_find_by_num(request->packet->vps, PW_REALM, 0, TAG_ANY) != NULL ) {
		RDEBUG2("Request already has destination realm set.  Ignoring");
		return RLM_MODULE_NOOP;
	}

	/*
	 *	We will be modifing this later, so we want our own copy
	 *	of it.
	 */
	namebuf = talloc_typed_strdup(request,  request->username->vp_strvalue);
	username = namebuf;

	switch (inst->format) {
	case REALM_FORMAT_SUFFIX:
		RDEBUG2("Checking for suffix after \"%c\"", inst->delim[0]);
		ptr = strrchr(username, inst->delim[0]);
		if (ptr) {
			*ptr = '\0';
			realmname = ptr + 1;
		}
		break;

	case REALM_FORMAT_PREFIX:
		RDEBUG2("Checking for prefix before \"%c\"", inst->delim[0]);
		ptr = strchr(username, inst->delim[0]);
		if (ptr) {
			*ptr = '\0';
			ptr++;
			realmname = username;
			username = ptr;
		}
		break;

	default:
		realmname = NULL;
		break;
	}

	/*
	 *	Print out excruciatingly descriptive debugging messages
	 *	for the people who find it too difficult to think about
	 *	what's going on.
	 */
	if (realmname) {
		RDEBUG2("Looking up realm \"%s\" for User-Name = \"%s\"",
		       realmname, request->username->vp_strvalue);
	} else {
		if (inst->ignore_null ) {
			RDEBUG2("No '%c' in User-Name = \"%s\", skipping NULL due to config.",
			inst->delim[0], request->username->vp_strvalue);
			talloc_free(namebuf);
			return RLM_MODULE_NOOP;
		}
		RDEBUG2("No '%c' in User-Name = \"%s\", looking up realm NULL",
			inst->delim[0], request->username->vp_strvalue);
	}

	/*
	 *	Allow DEFAULT realms unless told not to.
	 */
	realm = realm_find(realmname);

#ifdef HAVE_TRUST_ROUTER_TR_DH_H
	/*
	 *	Try querying for the dynamic realm.
	 */
	if (!realm) {
		if (inst->trust_router) {
			realm = tr_query_realm(request, realmname, inst->default_community, inst->rp_realm,
					       inst->trust_router, inst->tr_port);
		}
		else {
			RDEBUG2("No trust router configured, skipping dynamic realm lookup");
		}
	}
#endif

	if (!realm) {
		RDEBUG2("No such realm \"%s\"", (!realmname) ? "NULL" : realmname);
		talloc_free(namebuf);
		return RLM_MODULE_NOOP;
	}

	if (inst->ignore_default && (strcmp(realm->name, "DEFAULT")) == 0) {
		RDEBUG2("Found DEFAULT, but skipping due to config");
		talloc_free(namebuf);
		return RLM_MODULE_NOOP;
	}

	RDEBUG2("Found realm \"%s\"", realm->name);

	/*
	 *	If we've been told to strip the realm off, then do so.
	 */
	if (realm->strip_realm) {
		/*
		 *	Create the Stripped-User-Name attribute, if it
		 *	doesn't exist.
		 *
		 */
		if (request->username->da->attr != PW_STRIPPED_USER_NAME) {
			vp = radius_pair_create(request->packet, &request->packet->vps,
					       PW_STRIPPED_USER_NAME, 0);
			RDEBUG2("Adding Stripped-User-Name = \"%s\"", username);
		} else {
			vp = request->username;
			RDEBUG2("Setting Stripped-User-Name = \"%s\"", username);
		}

		fr_pair_value_strcpy(vp, username);
		request->username = vp;
	}

	/*
	 *	Add the realm name to the request.
	 *	If the realm is a regex, the use the realm as entered
	 *	by the user.  Otherwise, use the configured realm name,
	 *	as realm name comparison is case insensitive.  We want
	 *	to use the configured name, rather than what the user
	 *	entered.
	 */
	if (realm->name[0] != '~') realmname = realm->name;

	/*
	 *	A NULL realmname is allowed.
	 */
	if (realmname) {
		pair_make_request("Realm", realmname, T_OP_EQ);
		RDEBUG2("Adding Realm = \"%s\"", realmname);
	}

	talloc_free(namebuf);
	username = NULL;

	/*
	 *	Figure out what to do with the request.
	 */
	switch (request->packet->code) {
	default:
		RDEBUG2("Unknown packet code %d\n",
		       request->packet->code);
		return RLM_MODULE_NOOP;

		/*
		 *	Perhaps accounting proxying was turned off.
		 */
	case PW_CODE_ACCOUNTING_REQUEST:
		if (!realm->acct_pool) {
			RDEBUG2("Accounting realm is LOCAL");
			return RLM_MODULE_OK;
		}
		break;

		/*
		 *	Perhaps authentication proxying was turned off.
		 */
	case PW_CODE_ACCESS_REQUEST:
		if (!realm->auth_pool) {
			RDEBUG2("Authentication realm is LOCAL");
			return RLM_MODULE_OK;
		}
		break;
	}

#ifdef WITH_PROXY
	RDEBUG2("Proxying request from user %s to realm %s",
	       request->username->vp_strvalue, realm->name);

	/*
	 *	Skip additional checks if it's not an accounting
	 *	request.
	 */
	if (request->packet->code != PW_CODE_ACCOUNTING_REQUEST) {
		*returnrealm = realm;
		return RLM_MODULE_UPDATED;
	}

	/*
	 *	FIXME: Each server should have a unique server key,
	 *	and put it in the accounting packet.  Every server
	 *	should know about the keys, and NOT proxy requests to
	 *	a server with key X if the packet already contains key
	 *	X.
	 */

	/*
	 *      If this request has arrived from another freeradius server
	 *      that has already proxied the request, we don't need to do
	 *      it again.
	 */
	vp = fr_pair_find_by_num(request->packet->vps, PW_FREERADIUS_PROXIED_TO, 0, TAG_ANY);
	if (vp && (request->packet->src_ipaddr.af == AF_INET)) {
		int i;
		fr_ipaddr_t my_ipaddr;

		my_ipaddr.af = AF_INET;
		my_ipaddr.prefix = 32;
		my_ipaddr.ipaddr.ip4addr.s_addr = vp->vp_ipaddr;

		/*
		 *	Loop over the home accounting servers for this
		 *	realm.  If one of them has the same IP as the
		 *	FreeRADIUS-Proxied-To attribute, then the
		 *	packet has already been sent there.  Don't
		 *	send it there again.
		 */
		for (i = 0; i < realm->acct_pool->num_home_servers; i++) {
			if (realm->acct_pool->servers[i]->ipaddr.af == AF_UNSPEC) continue;

			if (fr_ipaddr_cmp(&realm->acct_pool->servers[i]->ipaddr, &my_ipaddr) == 0) {
				RDEBUG2("Suppressing proxy due to FreeRADIUS-Proxied-To");
				return RLM_MODULE_OK;
			}
		}

		/*
		 *	See detail_recv() in src/main/listen.c for the
		 *	additional checks.
		 */
#ifdef WITH_DETAIL
	} else if ((request->listener->type == RAD_LISTEN_DETAIL) &&
		   !fr_inaddr_any(&request->packet->src_ipaddr)) {
		int i;

		/*
		 *	Loop over the home accounting servers for this
		 *	realm.  If one of them has the same IP as the
		 *	FreeRADIUS-Proxied-To attribute, then the
		 *	packet has already been sent there.  Don't
		 *	send it there again.
		 */
		for (i = 0; i < realm->acct_pool->num_home_servers; i++) {
			if (realm->acct_pool->servers[i]->ipaddr.af == AF_UNSPEC) continue;

			if ((fr_ipaddr_cmp(&realm->acct_pool->servers[i]->ipaddr,
					     &request->packet->src_ipaddr) == 0) &&
			    (realm->acct_pool->servers[i]->port == request->packet->src_port)) {
				RDEBUG2("Suppressing proxy because packet was already sent to a server in that realm");
				return RLM_MODULE_OK;
			}
		}
#endif	/* WITH_DETAIL */
	}
#endif	/* WITH_PROXY */

	/*
	 *	We got this far, which means we have a realm, set returnrealm
	 */
	*returnrealm = realm;

	return RLM_MODULE_UPDATED;
}

/*
 *  Perform the realm module instantiation.  Configuration info is
 *  stored in *instance for later use.
 */

static int mod_instantiate(CONF_SECTION *conf, void *instance)
{
	struct rlm_realm_t *inst = instance;

	if (strcasecmp(inst->format_string, "suffix") == 0) {
	     inst->format = REALM_FORMAT_SUFFIX;

	} else if (strcasecmp(inst->format_string, "prefix") == 0) {
	     inst->format = REALM_FORMAT_PREFIX;

	} else {
		cf_log_err_cs(conf, "Invalid value \"%s\" for format",
			      inst->format_string);
	     return -1;
	}

	if (cf_new_escape && (strcmp(inst->delim, "\\\\") == 0)) {
		/* it's OK */
	} else

	if (strlen(inst->delim) != 1) {
		cf_log_err_cs(conf, "Invalid value \"%s\" for delimiter",
			      inst->delim);
	     return -1;
	}

#ifdef HAVE_TRUST_ROUTER_TR_DH_H
	/* initialize the trust router integration code */
	if (strcmp(inst->trust_router, "none") != 0) {
		if (!tr_init(inst->rekey_enabled, inst->realm_lifetime)) return -1;
	} else {
		rad_const_free(inst->trust_router);
		inst->trust_router = NULL;
	}
#endif

	return 0;
}


/*
 *  Examine a request for a username with an realm, and if it
 *  corresponds to something in the realms file, set that realm as
 *  Proxy-To.
 *
 *  This should very nearly duplicate the old proxy_send() code
 */
static rlm_rcode_t CC_HINT(nonnull) mod_authorize(void *instance, REQUEST *request)
{
	rlm_rcode_t rcode;
	REALM *realm;

	/*
	 *	Check if we've got to proxy the request.
	 *	If not, return without adding a Proxy-To-Realm
	 *	attribute.
	 */
	rcode = check_for_realm(instance, request, &realm);
	if (rcode != RLM_MODULE_UPDATED) return rcode;
	if (!realm) return RLM_MODULE_NOOP;

	/*
	 *	Maybe add a Proxy-To-Realm attribute to the request.
	 */
	RDEBUG2("Preparing to proxy authentication request to realm \"%s\"\n",
	       realm->name);
	pair_make_config("Proxy-To-Realm", realm->name, T_OP_EQ);

	return RLM_MODULE_UPDATED; /* try the next module */
}

/*
 * This does the exact same thing as the mod_authorize, it's just called
 * differently.
 */
static rlm_rcode_t CC_HINT(nonnull) mod_preacct(void *instance, REQUEST *request)
{
	int rcode;
	REALM *realm;

	if (!request->username) {
		return RLM_MODULE_NOOP;
	}

	/*
	 *	Check if we've got to proxy the request.
	 *	If not, return without adding a Proxy-To-Realm
	 *	attribute.
	 */
	rcode = check_for_realm(instance, request, &realm);
	if (rcode != RLM_MODULE_UPDATED) return rcode;
	if (!realm) return RLM_MODULE_NOOP;

	/*
	 *	Maybe add a Proxy-To-Realm attribute to the request.
	 */
	RDEBUG2("Preparing to proxy accounting request to realm \"%s\"\n",
	       realm->name);
	pair_make_config("Proxy-To-Realm", realm->name, T_OP_EQ);

	return RLM_MODULE_UPDATED; /* try the next module */
}

#ifdef WITH_COA
/*
 *	CoA realms via Operator-Name.  Because the realm isn't in a
 *	User-Name, concepts like "prefix" and "suffix' don't matter.
 */
static rlm_rcode_t mod_realm_recv_coa(UNUSED void *instance, REQUEST *request)
{
	VALUE_PAIR *vp;
	REALM *realm;

	if (fr_pair_find_by_num(request->packet->vps, PW_REALM, 0, TAG_ANY) != NULL) {
		RDEBUG2("Request already has destination realm set.  Ignoring");
		return RLM_MODULE_NOOP;
	}

	vp = fr_pair_find_by_num(request->packet->vps, PW_OPERATOR_NAME, 0, TAG_ANY);
	if (!vp) return RLM_MODULE_NOOP;

	/*
	 *	Catch the case of broken dictionaries.
	 */
	if (vp->da->type != PW_TYPE_STRING) return RLM_MODULE_NOOP;

	/*
	 *	The string is too short.
	 */
	if (vp->vp_length == 1) return RLM_MODULE_NOOP;

	/*
	 *	'1' means "the rest of the string is a realm"
	 */
	if (vp->vp_strvalue[0] != '1') return RLM_MODULE_NOOP;

	realm = realm_find(vp->vp_strvalue + 1);
	if (!realm) return RLM_MODULE_NOTFOUND;

	if (!realm->coa_pool) {
		RDEBUG2("CoA realm is LOCAL");
		return RLM_MODULE_OK;
	}

	/*
	 *	Maybe add a Proxy-To-Realm attribute to the request.
	 */
	RDEBUG2("Preparing to proxy authentication request to realm \"%s\"\n",
	       realm->name);
	pair_make_config("Proxy-To-Realm", realm->name, T_OP_EQ);

	return RLM_MODULE_UPDATED; /* try the next module */
}
#endif

/* globally exported name */
extern module_t rlm_realm;
module_t rlm_realm = {
	.magic		= RLM_MODULE_INIT,
	.name		= "realm",
	.type		= RLM_TYPE_HUP_SAFE,
	.inst_size	= sizeof(struct rlm_realm_t),
	.config		= module_config,
	.instantiate	= mod_instantiate,
	.methods = {
		[MOD_AUTHORIZE]		= mod_authorize,
		[MOD_PREACCT]		= mod_preacct,
#ifdef WITH_COA
		[MOD_RECV_COA]		= mod_realm_recv_coa
#endif
	},
};

